src/MiniCover/HitServices/HitService.cs

```diff
@@ -0,0 +1,90 @@
+﻿using MiniCover.HitServices;
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Threading;
+
+namespace MiniCover
+{
+    public static class HitService
+    {
+        private static ConcurrentDictionary<string, ConcurrentDictionary<int, Hit>> files = new ConcurrentDictionary<string, ConcurrentDictionary<int, Hit>>();
+
+        public static void Init(string fileName)
+        {
+            files.GetOrAdd(fileName, (f) => new ConcurrentDictionary<int, Hit>());
+        }
+
+        public static MethodContext EnterMethod(string fileName)
+        {
+            var file = files.GetOrAdd(fileName, f => new ConcurrentDictionary<int, Hit>());
+
+            return new MethodContext(file);
```

Comment:
I was doing optimizations also, I got similar results as you did, but I took a slightly different approach.
I changed the instrumentation to call EnterMethod at the beginning of every method, and use the MethodContext to record hits on instructions.
That allows us to compute some things only once per method, instead of doing it for reach instruction.
And also allow us to know when the method exited. I'm using that to clear the AsyncLocal value.

It seems your optimization with AsyncLocal is very similar, but I didn't see where it clears the current value. Are you getting all lines with same test method information as a side effect?